### PR TITLE
[SLP] Compute a shuffle mask for getGatherCost

### PR DESCRIPTION
**Pull request by preames**
This is the second of a series of small patches to compute shuffle masks for the couple of cases where we call getShuffleCost without one. My goal is to add an invariant that all calls to getShuffleCost for fixed length vectors have a mask.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -10516,6 +10516,7 @@ InstructionCost BoUpSLP::getGatherCost(ArrayRef<Value *> VL,
           TTI->getVectorInstrCost(Instruction::InsertElement, VecTy, CostKind,
                                   I, Constant::getNullValue(VecTy), V);
   };
+  SmallVector<int> ShuffleMask(VL.size(), -1);
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10519]:
```suggestion
  SmallVector<int> ShuffleMask(VL.size(), PoisonMaskElem);
```

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -10526,6 +10527,7 @@ InstructionCost BoUpSLP::getGatherCost(ArrayRef<Value *> VL,
     if (!UniqueElements.insert(V).second) {
       DuplicateNonConst = true;
       ShuffledElements.setBit(I);
+      ShuffleMask[I] = I;
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10530]: This does not look correct. You generated identity mask here, which is not. Need to record the position of the first unique value and then use it here as index.

Comment (preames) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10530]: Clearly it is time for me to stop for the day.  I wrote the correct code here - and then managed to commit the wrong version.  Oops.  Will push a corrected version tomorrow.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -10516,27 +10516,34 @@ InstructionCost BoUpSLP::getGatherCost(ArrayRef<Value *> VL,
           TTI->getVectorInstrCost(Instruction::InsertElement, VecTy, CostKind,
                                   I, Constant::getNullValue(VecTy), V);
   };
+  SmallVector<int> ShuffleMask(VL.size(), PoisonMaskElem);
   for (unsigned I = 0, E = VL.size(); I < E; ++I) {
     Value *V = VL[I];
     // No need to shuffle duplicates for constants.
     if ((ForPoisonSrc && isConstant(V)) || isa<UndefValue>(V)) {
       ShuffledElements.setBit(I);
+      ShuffleMask[I] = I;
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10525]:
```suggestion
      ShuffleMask[I] = isa<PoisonValue>(V) ? PoisonMaskElem : I;
```

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 4]
```diff
@@ -10516,27 +10516,34 @@ InstructionCost BoUpSLP::getGatherCost(ArrayRef<Value *> VL,
...
       continue;
     }
-    if (!UniqueElements.insert(V).second) {
-      DuplicateNonConst = true;
-      ShuffledElements.setBit(I);
+
+    if (!UniqueElements.count(V)) {
+      EstimateInsertCost(I, V);
+      UniqueElements[V] = I;
+      ShuffleMask[I] = I;
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10532]:
```suggestion
    auto Res = UniqueElements.try_emplace(V, I);
    if (Res.second) {
      EstimateInsertCost(I, V);
      ShuffleMask[I] = I;
```

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 5]
```diff
@@ -10516,27 +10516,34 @@ InstructionCost BoUpSLP::getGatherCost(ArrayRef<Value *> VL,
...
       continue;
     }
-    EstimateInsertCost(I, V);
+
+    DuplicateNonConst = true;
+    ShuffledElements.setBit(I);
+    ShuffleMask[I] = UniqueElements[V];
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10538]:
```suggestion
    ShuffleMask[I] = Res.first->second;
```

**Review by alexey-bataev — COMMENTED** [review]
Can you add a test?

**Comment by preames** [comment]
> Can you add a test?

I have no idea how to construct an SLP test which exercises this.

**Review by alexey-bataev — APPROVED** [review]
LG

**Pull request merged**